packages/chain-helpers/src/blockchain/Blockchain.ts

```diff
@@ -174,19 +176,47 @@ export const dispatchTx = submitSignedTx
  * @param signer The {@link KeyringPair} used to sign the tx.
  * @param opts Additional options.
  * @param opts.tip Optional amount of Femto-KILT to tip the validator.
+ * @param opts.checkMetadata Boolean flag indicated whether to verify the metadata hash upon tx submission.
  * @returns A signed {@link SubmittableExtrinsic}.
  */
 export async function signTx(
   tx: SubmittableExtrinsic,
   signer: KeyringPair | TransactionSigner,
-  { tip }: { tip?: AnyNumber } = {}
+  { tip, checkMetadata }: { tip?: AnyNumber; checkMetadata?: boolean } = {}
 ): Promise<SubmittableExtrinsic> {
+  const signOptions = await (async (): Promise<Partial<SignerOptions>> => {
+    if (!checkMetadata) {
+      return {
+        tip,
+      }
+    }
+    // If `checkMetadata` is enabled, include that when signing the tx.
+    const api = ConfigService.get('api')
+    const metadata = await api.call.metadata.metadataAtVersion(15)
+    const { specName, specVersion } = api.runtimeVersion
+    const merkleInfo = {
+      base58Prefix: api.consts.system.ss58Prefix.toNumber(),
+      decimals: api.registry.chainDecimals[0],
+      specName: specName.toString(),
+      specVersion: specVersion.toNumber(),
+      tokenSymbol: api.registry.chainTokens[0],
+    }
+    const merkleizedMetadata = merkleizeMetadata(metadata.toHex(), merkleInfo)
+    const metadataHash = u8aToHex(merkleizedMetadata.digest())
```

Comment:
the metadata hash calculation seems pretty computation heavy, we could maybe cache this in a map from genesis hash and runtime version to metadataHash?

Reply:
When would you populate the cache? First time we do this? Where would we store such a map? In the config service?

Reply:
I would keep it local to this package, and only populate the cache when a hash is requested but a cache miss occurs

Reply:
I gave it a try. What do you think? I'm using the `ConfigService` map (I could not think of anything else), and using (genesis hash, spec name, and spec version) as the key and the metadata hash as the value: https://github.com/KILTprotocol/sdk-js/pull/924/commits/dfd17af1f20716866adb67d07eb1d937a01f44e8

Reply:
https://github.com/KILTprotocol/sdk-js/pull/924/commits/22c3af8d938ed0f7c3acb1ec5f551731ccf5774b